File: 0430-flatten-a-multilevel-doubly-linked-list/0430-flatten-a-multilevel-doubly-linked-list.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution(object):
    def flatten(self, head):
        """
        :type head: Node
        :rtype: Node
        """
        if not head : return head 
        cur = head 
        stack = []
        while cur :
            if cur.child :
                if cur.next :
                    stack.append(cur.next)
                cur.next = cur.child 
                cur.child.prev = cur
                cur.child = None

            if not cur.next  and stack :
                node = stack.pop()
                cur.next = node 
                if node : node.prev = cur 
            cur = cur.next 
        return head 
        

File: 0430-flatten-a-multilevel-doubly-linked-list/0430-flatten-a-multilevel-doubly-linked-list.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;
};
*/

class Solution {
    public Node helper(Node head){
        Node last ,temp = head,prev = null ;
        while (temp != null){
            prev = temp ;
            if(temp.child != null){
                last = helper(temp.child) ;
                last.next = temp.next ;
                temp.next = temp.child ;
                temp.child = null ;
                temp.next.prev = temp ;
                if(last.next != null) last.next.prev = last ;


            }
            temp = temp.next ; 
        }
        return prev ;
    }
    public Node flatten(Node head) {
        helper(head) ;
        return head ;
        
        
    }
}